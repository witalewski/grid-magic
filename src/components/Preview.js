import React, { useEffect, useRef } from 'react';
import { string } from 'prop-types';
import { connect } from 'react-redux';

export const Preview = ({ image }) => {
  const canvas = useRef(null);

  const width = (1080 * 3 + 2) / 4;
  const height = 1080 / 4;

  useEffect(() => {
    if (image && canvas.current) {
      const ctx = canvas.current.getContext('2d');
      const img = new Image();
      img.onload = () => {
        ctx.drawImage(img, 0, 0, width, height);
      };
      img.src = image;
    }
  });

  return image && <canvas ref={canvas} alt="Preview" width={width} height={height} />;
};

Preview.propTypes = {
  image: string,
};

Preview.defaultProps = {
  image: '',
};

const mapStateToProps = state => ({
  image: state.image.image,
});

export const ConnectedPreview = connect(mapStateToProps)(Preview);
