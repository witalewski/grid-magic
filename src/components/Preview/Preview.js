import React, { Component } from 'react';
import { object } from 'prop-types';
import { inject, observer } from 'mobx-react';

class Preview extends Component {
  static propTypes = {
    previewCanvas: object,
  };
  static defaultProps = {
    previewCanvas: null,
  };

  render() {
    return (
      <img
        alt="Preview"
        className="preview-image"
        src={
          this.props.previewCanvas &&
          this.props.previewCanvas.toDataURL('image/png')
        }
      />
    );
  }
}

export { Preview };
export default inject(({ appState }) => ({
  previewCanvas: appState.previewCanvas,
}))(observer(Preview));
