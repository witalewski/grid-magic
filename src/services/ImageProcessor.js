import { Promise } from 'rsvp';

export class ImageProcessor {
  createCanvas = (width, height) => {
    const canvas = document.createElement('canvas');
    canvas.width = width;
    canvas.height = height;
    return canvas;
  };

  getPlaceholderCanvas = (width, height, tileSize, gapSize) => {
    const canvas = this.createCanvas(width, height);
    const context = canvas.getContext('2d');
    context.fillStyle = 'lightgray';
    context.fillRect(0, 0, width, height);
    context.fillStyle = 'white';
    context.fillRect(tileSize + gapSize, 0, gapSize, tileSize);
    context.fillRect(tileSize * 2 + gapSize * 2, 0, gapSize, tileSize);
    return canvas;
  };

  getCanvasFromFile = async (file, width, height, tileSize, gapSize) =>
    new Promise(resolve => {
      const fileReader = new FileReader();

      fileReader.onload = event => {
        const imageData = event.target.result;
        const img = new Image();

        img.onload = ({ target }) => {
          const canvas = this.createCanvas(width, height);
          const context = canvas.getContext('2d');
          context.fillStyle = 'white';
          context.fillRect(0, 0, width, height);
          let targetWidth;
          let targetHeight;
          let xOffset;
          let yOffset;
          const imgProportions = target.width / target.height;
          const canvasProportions = width / height;
          if (imgProportions > canvasProportions) {
            targetWidth = width;
            targetHeight = width / imgProportions;
            xOffset = 0;
            yOffset = (height - targetHeight) / 2;
          } else {
            targetWidth = height * imgProportions;
            targetHeight = height;
            xOffset = (width - targetWidth) / 2;
            yOffset = 0;
          }
          context.drawImage(
            target,
            xOffset,
            yOffset,
            targetWidth,
            targetHeight
          );
          context.fillRect(tileSize + gapSize, 0, gapSize, tileSize);
          context.fillRect(tileSize * 2 + gapSize * 2, 0, gapSize, tileSize);

          resolve(canvas);
        };
        
        img.src = imageData;
      };

      fileReader.readAsDataURL(file);
    });

  saveBase64AsFile = (base64, fileName) => {
    const link = document.createElement('a');
    link.setAttribute('href', base64);
    link.setAttribute('download', fileName);
    link.click();
  };

  downloadImages = (sourceCanvas, tileSize, gapSize) => {
    const exportCanvas = this.createCanvas(tileSize, tileSize);
    const exportContext = exportCanvas.getContext('2d');
    [tileSize * 2 + gapSize * 3, tileSize + gapSize * 2, 0].forEach((x, i) => {
      exportContext.drawImage(
        sourceCanvas,
        x,
        0,
        tileSize,
        tileSize,
        0,
        0,
        tileSize,
        tileSize
      );

      const data = exportCanvas.toDataURL('image/png');
      console.log(data);
      this.saveBase64AsFile(data, `export-${3-i}.png`);
    });
  };
}
