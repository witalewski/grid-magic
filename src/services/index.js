export { ImageProcessor } from './ImageProcessor/ImageProcessor';
