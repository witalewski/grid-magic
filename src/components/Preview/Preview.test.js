import React from 'react';
import { Preview } from './Preview';
import { ImageProcessor } from '../../services';
import ShallowRenderer from 'react-test-renderer/shallow';

describe('Preview', () => {
  it('displays preview', () => {
    const renderer = new ShallowRenderer();
    const imageProcessor = new ImageProcessor();
    renderer.render(
      <Preview canvas={imageProcessor.getPlaceholderCanvas(100, 100, 100, 0)} />
    );
    const result = renderer.getRenderOutput();

    expect(result).toMatchSnapshot();
  });
});
