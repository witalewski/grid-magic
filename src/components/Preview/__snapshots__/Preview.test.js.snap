// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview displays preview 1`] = `
<img
  alt="Preview"
  className="preview-image"
  onDrop={[Function]}
  src={null}
/>
`;
